modules/onewire/onewire.c

```diff
@@ -0,0 +1,418 @@
+    /* Restore the search state */
+    for (i = 0; i < ROM_DATA_SIZE_BYTES; i++) {
+        OneWireStruct->ROM_NO[i] = rom_backup[i];
+    }
+    OneWireStruct->LastDiscrepancy = ld_backup;
+    OneWireStruct->LastDeviceFlag = ldf_backup;
+    OneWireStruct->LastFamilyDiscrepancy = lfd_backup;
+
+    /* Return the result of the verify */
+    return rslt;
+}
+
+uint8_t OneWire_TargetSetup(OneWire_t* OneWireStruct, uint8_t family_code) {
+    if (!OneWireStruct) return 0;
+    uint8_t i;
+
+    /* Set the search state to find SearchFamily type devices */
+    OneWireStruct->ROM_NO[0] = family_code;
+    for (i = 1; i < ROM_DATA_SIZE_BYTES; i++) {
+        OneWireStruct->ROM_NO[i] = 0;
+    }
+
+    OneWireStruct->LastDiscrepancy = ROM_DATA_SIZE_BITS;
+    OneWireStruct->LastFamilyDiscrepancy = 0;
+    OneWireStruct->LastDeviceFlag = 0;
+    return 1;
+}
+
+uint8_t OneWire_FamilySkipSetup(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return 0;
+    /* Set the Last discrepancy to last family discrepancy */
+    OneWireStruct->LastDiscrepancy = OneWireStruct->LastFamilyDiscrepancy;
+    OneWireStruct->LastFamilyDiscrepancy = 0;
+
+    /* Check for end of list */
+    if (OneWireStruct->LastDiscrepancy == 0) {
+        OneWireStruct->LastDeviceFlag = 1;
+    }
+    return 1;
+}
+
+char OneWire_GetROM(OneWire_t* OneWireStruct, uint8_t index) {
+    if (!OneWireStruct)                           return -1;
+    if (index > 0 || index < ROM_DATA_SIZE_BYTES) return -1;
+    return OneWireStruct->ROM_NO[index];
+}
+
+uint8_t OneWire_Select(OneWire_t* OneWireStruct, uint8_t* addr) {
+    if (!OneWireStruct) return 0;
+    if (!addr)          return 0;
+    uint8_t i;
+    OneWire_WriteByte(OneWireStruct, ONEWIRE_CMD_MATCHROM);
+    
+    for (i = 0; i < ROM_DATA_SIZE_BYTES; i++) {
+        OneWire_WriteByte(OneWireStruct, *(addr + i));
+    }
+    return 1;
+}
+
+uint8_t OneWire_SelectWithPointer(OneWire_t* OneWireStruct, uint8_t *ROM) {
+    if (!OneWireStruct) return 0;
+    if (!ROM)           return 0;
+    uint8_t i;
+    OneWire_WriteByte(OneWireStruct, ONEWIRE_CMD_MATCHROM);
+    
+    for (i = 0; i < ROM_DATA_SIZE_BYTES; i++) {
+        OneWire_WriteByte(OneWireStruct, *(ROM + i));
+    }   
+    return 1;
+}
+
+uint8_t OneWire_GetFullROM(OneWire_t* OneWireStruct, uint8_t *firstIndex) {
+    if (!OneWireStruct) return 0;
+    if (!firstIndex)    return 0;
+    uint8_t i;
+    for (i = 0; i < ROM_DATA_SIZE_BYTES; i++) {
+        *(firstIndex + i) = OneWireStruct->ROM_NO[i];
+    }
+    return 1;
+}
+
+char OneWire_CRC8(uint8_t *addr, uint8_t len) {
+    if (!addr) return -1;
+    uint8_t crc = 0, inbyte, i, mix;
+    while (len--) {
+        inbyte = *addr++;
+        for (i = ROM_DATA_SIZE_BYTES; i; i--) {
+            mix = (crc ^ inbyte) & 0x01;
+            crc >>= 1;
+            if (mix) {
+                crc ^= 0x8C;
+            }
+            inbyte >>= 1;
+        }
+    }
+    
+    /* Return calculated CRC */
+    return crc;
+}

```

Comment:
Something to consider.  You can speed up an 8-bit CRC calculation by using either 1) a lookup table or 2) the CRC peripheral on the MCU.

I suggest this because a lot of the OneWire calls are blocking calls so we should be looking to save MIPS where ever we can.